_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+   kani::assume(queue.is_layout_valid(&mem));
+
+    let needs_notification = queue.prepare_kick(&mem);
+
+    if !queue.uses_notif_suppression {
+        // The specification here says
+        // After the device writes a descriptor index into the used ring:
+        // – If flags is 1, the device SHOULD NOT send a notification.
+        // – If flags is 0, the device MUST send a notification
+        // flags is the first field in the avail_ring, which we completely ignore. We
+        // always send a notification, and as there only is a SHOULD NOT, that is okay
+        assert!(needs_notification);
+    } else {
+        // next_used - 1 is where the previous descriptor was placed.
+       // queue.used_event(&mem) reads from the memory location at which the guest
+       // stores the index for which it wants to receive the next notification
+        if queue.used_event(&mem) == queue.next_used - Wrapping(1) && queue.num_added.0 > 0 {
+            // If the idx field in the used ring (which determined where that descriptor index
+            // was placed) was equal to used_event, the device MUST send a
+            // notification.
+            assert!(needs_notification)
+        }
+
+        // The other case is handled by a "SHOULD NOT send a notification" in the spec.
+        // So we do not care
+    }
+}
+```
+
+Beyond these specification conformance harnesses, we also have standard “absense of panics” harnesses, which lead us to discover an issue in our code the validates the in-memory layout of VirtIO queues. A guest should trigger a panic in firecracker by placing the starting address for a VirtIO queue component into the MMIO gap.
+
+## Conclusion
+
+Thanks to Kani, the Firecracker team was able to test critical areas of code that were intractable to traditional methods. These include our noisy-neighbor mitigation, a rate limiter, where interactions with the system clock resulted in traditional testing being unreliable, as well as our VirtIO stack, where the interaction with guest memory lead to a state space impossible to cover by other means. 
+
+We found 5 bugs in our rate limiter implementation, the most significant one a rounding error that allowed guests to exceed their prescribed I/O bandwidth by up to 0.1% in some cases. Additionally, we found one bug in our VirtIO stack, where a malicious guest could set up a virtio queue that partially overlapped with the MMIO memory region, resulting in Firecracker crashing on boot. Finally, the added debug assertions added to the code under test allowed us to identify a handful of unit tests which were not set up correctly. These have also been fixed. 
+
+All in all, Kani proof harnesses has proven a valuable defense-in-depth measure for Firecracker, nicely complementing our existing testing infrastructure. We plan to continue our investment in these harnesses as we develop new Firecracker features, to ensure consistently high security standards. 
+
```

Comment:
You also need to include author bio!

Reply:
Haha, okay, I'll put it back. I removed it after adding my name in the introduction because there was really nothing interesting to say lol